lib/log.h

```diff
@@ -71,16 +71,36 @@ extern enum metal_log_level metal_get_log_level(void);
 extern void metal_default_log_handler(enum metal_log_level level,
 				      const char *format, ...);
 
+/**
+ * @brief	used by the metal_log() macro to update the format string
+ * @fmt	format string passed from the metal_log() macro
+ *
+ * If ML_FUNC_LINE is defined this macro generates a unified format
+ * string for metal_log() and its convenience ml_*() macros, i.e. it
+ * adds function-name:line-number prefix to all log messages.
+ */
+#if defined(ML_FUNC_LINE)
+#define ml_fmt(fmt) "%s:%u " fmt, __func__, __LINE__
+#else	/* ML_FUNC_LINE */
+#define ml_fmt(fmt) fmt
+#endif	/* ML_FUNC_LINE */
+
 /**
  * Emit a log message if the log level permits.
  *
  * @param	level	Log level.
  * @param	...	Format string and arguments.
  */
-#define metal_log(level, ...)						       \
-	((level <= _metal.common.log_level && _metal.common.log_handler) \
-	       ? (void)_metal.common.log_handler(level, __VA_ARGS__)	       \
-	       : (void)0)
+#define metal_log(level, fmt, args ...)					\
```

Comment:
',' between `args `and `...`?

Reply:
Adding ',' would result in compilation errors, e.g.:
```
error: macro "metal_log" requires 4 arguments, but only 2 given
   | metal_log(METAL_LOG_DEBUG, "failed to allocate memory\n");
   |                                                         ^
```
Please, see "Macros with a Variable Number of Arguments" section in the [gcc manual](https://github.com/torvalds/linux/blob/master/Documentation/process/coding-style.rst).

Reply:
Ok, I see that you suppress space between `args` and  `...` in your last version

Reply:
Thank you.